lib/shopify_app/login_protection.rb

```diff
@@ -41,20 +41,25 @@ def redirect_to_login
         head :unauthorized
       else
         session[:return_to] = request.fullpath if request.get?
-        redirect_to main_or_engine_login_url(shop: params[:shop])
+        redirect_to login_url
       end
     end
 
     def close_session
       session[:shopify] = nil
       session[:shopify_domain] = nil
-      redirect_to main_or_engine_login_url(shop: params[:shop])
+      redirect_to login_url
     end
 
-    def main_or_engine_login_url(params = {})
-      main_app.login_url(params)
-    rescue NoMethodError
-      shopify_app.login_url(params)
+    def login_url
+      url = ShopifyApp.configuration.login_url
+
+      if params[:shop].present?
+        query = { shop: params[:shop] }.to_query
+        url = "#{url}?#{query}"
```

Comment:
That sounds like a mistake no? Consider the case where someone is accessing the app through a [admin link](https://help.shopify.com/partners/making-apps/app-admin-access):

1. The link will send them to `app.com?shop=SHOP&action=ACTION`
2. If they are not logged in, they'll be redirected to `app.com/login?shop=SHOP`.
3. Once logged in, they'll probably be redirect to `app.com`